src/extension.ts

```diff
@@ -1,106 +1,213 @@
+	}
+
+	async function didOpenTextDocument(document: TextDocument): Promise<void> {
+
+		const uri = document.uri;
+
+		let folder = workspace.getWorkspaceFolder(uri);
+		// Files outside a folder can't be handled.
+		if (!folder) {
+			return;
+		}
+		// If we have nested workspace folders we only start a language server on the outer most workspace folder.
+		folder = getOuterMostWorkspaceFolder(folder);
+		lastActiveFolder = folder;
+		let folderURI = folder.uri.toString();
+
+		// We are only interested in .cf files
+		if (document.languageId !== 'inmanta' || (document.uri.scheme !== 'file')) {
+			return;
 		}
-	}));
 
 
-	// Start the language server if enabled in the workspace configuration
-	const enable: boolean = workspace.getConfiguration('inmanta').ls.enabled;
-	if (enable) {
-		await languageserver.startOrRestartLS(true);
-	}
+		if (!languageServers.has(folderURI)) {
+			/*
+				The document that was just opened is not living inside a folder that has a language server responsible for it.
+				We need to start a new language server for this folder. For a seamless user experience, we mimick the behaviour
+				of the pylance extension:
+				
+				- Case 1: a venv for this folder has already been selected in the past and persisted by vs code in the persistent
+				storage ==> we simply use this venv and start a new language server. see https://github.com/microsoft/vscode-python/wiki/Setting-descriptions#pythondefaultinterpreterpath)
 
+				- Case 2: this is a fresh folder with no pre-selected venv
+					* if a workspace-wide venv has been selected -> use this one
+					* use the default environment used by the python extension (https://code.visualstudio.com/docs/python/environments#_where-the-extension-looks-for-environments)
+				
+			*/
 
+			let newPath = pythonExtensionInstance.getPathForResource(folder.uri);
 
-}
 
-export async function deactivate(){
-	log("deactivate");
-	return languageserver.stopServerAndClient();
-}
+			// This check makes sure the active interpreter is part of a venv so we don't install anything in a global env.
+			// Inspired by: https://stackoverflow.com/questions/1871549/determine-if-python-is-running-inside-virtualenv/42580137#42580137
+			const script = "import sys\n" +
+				"real_prefix = getattr(sys, 'real_prefix', None)\n" +
+				"base_prefix = getattr(sys, 'base_prefix', sys.prefix)\n" +
+				"running_in_virtualenv = (base_prefix or real_prefix) != sys.prefix\n" +
+				"if not running_in_virtualenv:\n" +
+				"  sys.exit(1)";
+
+			let spawnResult = cp.spawnSync(newPath, ["-c", script]);
+			const stdout = spawnResult.stdout.toString();
+			if (spawnResult.status === 1) {
+				const response = await window.showErrorMessage(`The active python interpreter is not part of a virtual environment.`, 'Select interpreter');
+				if (response === 'Select interpreter') {
+					await commands.executeCommand('python.setInterpreter');
+				}
+				else {
+					return;
+				}
+			}
```

Comment:
I'm not sure I understand what you mean :p

Reply:
This check makes sure that the python interpreter chosen by the user is actually a venv ( and not a global interpreter like `/bin/python` for example) so that when we don't install module dependencies in the global context of the user. (`context` might not be the correct terminology :thinking: ).

I have now moved this check back to the canServerStart() method.